app/server/appsmith-server/src/main/java/com/appsmith/server/services/UserOrganizationServiceImpl.java

```diff
@@ -296,4 +293,66 @@ public UserOrganizationServiceImpl(SessionUserService sessionUserService,
+            userRole.setName(user.getName());
+            userRole.setRole(role);
+            newUserRoles.add(userRole);
+        }
+
+        if (newUserRoles.isEmpty()) {
+            // All the users being added to the organization already exist in the organization. Return without doing anything
+            // Because we are not erroring out here, this ensures that an email would be sent everytime a user is invited
+            // to an organization (whether or not the user is already part of the organization)
+            return Mono.just(organization);
+        }
+
+        // Add the users to the organization roles
+        userRoles.addAll(newUserRoles);
+
+        // Generate all the policies for Organization, Application, Page and Actions for the current user
+        Set<AclPermission> rolePermissions = role.getPermissions();
+        Map<String, Policy> orgPolicyMap = policyUtils.generatePolicyFromPermissionForMultipleUsers(rolePermissions, users);
+        Map<String, Policy> applicationPolicyMap = policyUtils.generateChildrenPoliciesFromOrganizationPolicies(orgPolicyMap, Application.class);
+        Map<String, Policy> datasourcePolicyMap = policyUtils.generateChildrenPoliciesFromOrganizationPolicies(orgPolicyMap, Datasource.class);
+        Map<String, Policy> pagePolicyMap = policyUtils.generatePagePoliciesFromApplicationPolicies(applicationPolicyMap);
+        Map<String, Policy> actionPolicyMap = policyUtils.generateActionPoliciesFromPagePolicies(pagePolicyMap);
+
+        //Now update the organization policies
+        Organization updatedOrganization = (Organization) policyUtils.addPoliciesToExistingObject(orgPolicyMap, organization);
+        updatedOrganization.setUserRoles(userRoles);
+
+        // Update the underlying application/page/action
+        Flux<Datasource> updatedDatasourcesFlux = policyUtils.updateWithNewPoliciesToDatasourcesByOrgId(updatedOrganization.getId(), datasourcePolicyMap, true);
+        Flux<Application> updatedApplicationsFlux = policyUtils.updateWithNewPoliciesToApplicationsByOrgId(updatedOrganization.getId(), applicationPolicyMap, true);
+        Flux<Page> updatedPagesFlux = updatedApplicationsFlux
+                .flatMap(application -> policyUtils.updateWithApplicationPermissionsToAllItsPages(application.getId(), pagePolicyMap, true));
+        Flux<Action> updatedActionsFlux = updatedPagesFlux
+                .flatMap(page -> policyUtils.updateWithPagePermissionsToAllItsActions(page.getId(), actionPolicyMap, true));
+
+        return Mono.zip(updatedDatasourcesFlux.collectList(), updatedActionsFlux.collectList(), Mono.just(updatedOrganization))
+                .flatMap(tuple -> {
+                    //By now all the datasources/applications/pages/actions have been updated. Just save the organization now
+                    Organization updatedOrgBeforeSave = tuple.getT3();
+                    return organizationRepository.save(updatedOrgBeforeSave);
+                });
```

Comment:
The third Mono in zip is not really necessary no? I think we can use the `updatedOrganization` directly.

```suggestion
        return Mono.zip(updatedDatasourcesFlux.collectList(), updatedActionsFlux.collectList())
                .flatMap(tuple -> {
                    //By now all the datasources/applications/pages/actions have been updated. Just save the organization now
                    return organizationRepository.save(updatedOrganization);
                });
```